jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/proxy/WebSocketProxy.java

```diff
@@ -0,0 +1,635 @@
+                callback.failed(failure);
+            else
+                server2Proxy.send(frame, sendCallback);
+        }
+
+        @Override
+        public void onError(Throwable failure, Callback callback)
+        {
+            System.err.println(toString() + " onError(): " + failure);
+
+            boolean failServer2Proxy;
+            synchronized (lock)
+            {
+                switch (state)
+                {
+                    case FAILED:
+                    case CLOSED:
+                        failServer2Proxy = false;
+                        break;
+
+                    default:
+                        state = State.FAILED;
+                        error = failure;
+                        failServer2Proxy = true;
+                        break;
+                }
+            }
+
+            if (failServer2Proxy)
+                server2Proxy.fail(failure,callback);
+            else
+                callback.failed(failure);
+        }
+
+        public void fail(Throwable failure, Callback callback)
+        {
+            System.err.println(toString() + " fail(): " + failure);
+
+            Callback sendCallback = null;
+            synchronized (lock)
+            {
+                switch (state)
+                {
+                    case OPEN:
+                        state = State.FAILED;
+                        sendCallback = Callback.from(callback, failure);
+                        break;
+
+                    case ICLOSED:
+                        state = State.FAILED;
+                        Callback doubleCallback = Callback.from(callback, closeCallback);
+                        sendCallback = Callback.from(doubleCallback, failure);
+                        break;
+
+                    default:
+                        state = State.FAILED;
+                        break;
+                }
+            }
+
+            if (sendCallback != null)
+                client.close(CloseStatus.SHUTDOWN, failure.getMessage(), sendCallback);
+            else
+                callback.failed(failure);
+        }
+
+        @Override
+        public void onClosed(CloseStatus closeStatus, Callback callback)
+        {
+            System.err.println(toString() + " onClosed(): " + closeStatus);
+            closed.countDown();
+            callback.succeeded();
+        }
+
+        public void send(Frame frame, Callback callback)
+        {
+            System.err.println(toString() + " send(): " + frame);
+
+            Callback sendCallback = callback;
+            Throwable failure = null;
+            synchronized (lock)
+            {
+                switch (state)
+                {
+                    case OPEN:
+                        if (frame.getOpCode() == OpCode.CLOSE)
+                            state = State.OCLOSED;
+                        break;
+
+                    case ICLOSED:
+                        if (frame.getOpCode() == OpCode.CLOSE)
```

Comment:
I think this is wrong?    When `client2proxy` receives the close frame it's onFrame method will change to `ICLOSED` and then send the close from using this method, so this will immediately go to `CLOSED`. 

I also don't like the double state change of grabbing lock and doing state handling in onFrame and then doing the same again in send.   I think we should be able to grab lock once, do the state change, do the action and then any subsequent state change needed should be done in the callback. `send(...)` is only called from `onFrame` so all the state change logic can be done there.